### ISSUE TITLE
typescript/flow typing

### ISSUE THREAD

**Issue by hkjeffchan** [jaydenseric/graphql-upload#70]
Thanks for this great library. Love it!

It will be great if you can add typing to typescript too.

@types/apollo-upload-server

**Comment by Slessi** [comment]
Using this in my project, (I don't use koa so I didn't add the type for that)

```typescript
declare module "apollo-upload-server" {
    import { Request, RequestHandler } from "express";
    import { GraphQLScalarType } from "graphql";

    export const GraphQLUpload: GraphQLScalarType;

    export type ApolloUploadOptions = {
        /**
         * Max allowed non-file multipart form field size in bytes; enough for your queries (default: 1 MB)
         */
        maxFieldSize?: number;
        /**
         * Max allowed file size in bytes (default: Infinity)
         */
        maxFileSize?: number;
        /**
         * Max allowed number of files (default: Infinity)
         */
        maxFiles?: number;
    };

    export function apolloUploadExpress(options?: ApolloUploadOptions): RequestHandler;
    export function processRequest(request: Request, options?: ApolloUploadOptions): Promise<any>;
}
```

**Comment by guatedude2** [comment]
+1

**Comment by mike-marcacci** [comment]
@hkjeffchan, @Slessi - I use flow in about a 9:1 ratio with ts, so I'm not completely sure what best practices are surrounding library defs. Is there any real benefit to keeping types in this repo vs using [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)?

**Comment by Slessi** [comment]
@mike-marcacci 

> tl;dr it really doesnt matter

The only difference is installing & maintaining one package versus two. Specifically `apollo-upload-server` and (if using DefinitelyTyped) `@types/apollo-upload-server`.

According to the [TS Handbook](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html), this library should probably use DefinitelyTyped.

> 1. bundling with your npm package, or
> 2. publishing to the `@types` organization on npm.
>
> If your package is written in TypeScript then the first approach is favored. Use the --declaration flag to generate declaration files. This way, your declarations and JavaScript always be in sync.
>
> If your package is not written in TypeScript then the second is the preferred approach.

At the same time, the core `apollographql` packages like [apollo-client](https://github.com/apollographql/apollo-client) are written in TS and as such bundle types with their packages, so inside the apollo ecosystem you *usually* don't have to.

**Comment by mike-marcacci** [comment]
Version `7.0.0` will be out very soon (probably today) which hopefully has the last big change to the API for a while. Once this is out, we should look at this again and get typing for both TypeScript and Flow.

I'm still not 100% sure about best practices RE using [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/)/[flow-typed](https://github.com/flow-typed/flow-typed) vs keeping type definitions here.

@jaydenseric do you have a preference? It feels cleaner to keep them here in this repo, but it also adds a (small but real) maintenance burden which could increase the barrier to entry for new contributions.

**Comment by mrdulin** [comment]
@Slessi thanks. Can you make a node module and put it on DefinitelyTyped?

**Comment by Slessi** [comment]
@mrdulin Types as found in my comment are probably incomplete.

At the very least its missing koa stuff, but I haven't looked at this library in a while so I don't know if anything else is missing. I wouldn't want to upload incomplete types to DT.

**Comment by jaydenseric** [comment]
I'm glad we held off on setting up DefinitelyTyped, since the `apollo-upload-server` package has been renamed `graphql-upload`, see [the v8.0.0 changelog entry](https://github.com/jaydenseric/graphql-upload/blob/master/changelog.md#800).

I'm not a TypeScript user and don't intend to oversee the publishing of types, but if anyone else does, please let us know here!

**Comment by eyexpo-ericl** [comment]
Not really skilled with typescript, so how would one get typescript typing? I don't understand what to do here .

**Comment by thomaseyexpo** [comment]
@eyexpo-ericl Create `graphql-upload.d.ts`

**Comment by mike-marcacci** [comment]
Types are now available by adding the `@types/graphql-upload` package.

**Comment by shoesel** [comment]
> Types are now available by adding the `@types/graphql-upload` package.

..not anymore? :(

**Comment by mike-marcacci** [comment]
🤦🏻‍♂️woops, I published types for most of the packages in this mini-ecosystem but forgot to open the PR for this one. This will be up shortly: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35276
